File: frontend/src/services/provinceService.ts
import api from './api';

export interface Province {
  id: number;
  code: string;
  name: string;
  nameEn: string;
  region: string;
  latitude: number;
  longitude: number;
}

const provinceService = {
  async getAllProvinces(): Promise<Province[]> {
    const response = await api.get('/provinces');
    return response.data.data;
  },

  async searchProvinces(keyword: string): Promise<Province[]> {
    const response = await api.get('/provinces/search', {
      params: { keyword }
    });
    return response.data.data;
  },

  async getProvincesByRegion(region: string): Promise<Province[]> {
    const response = await api.get(`/provinces/region/${region}`);
    return response.data.data;
  },

  async getProvinceByCode(code: string): Promise<Province> {
    const response = await api.get(`/provinces/code/${code}`);
    return response.data.data;
  },

  async getClosestProvince(latitude: number, longitude: number): Promise<Province | null> {
    const response = await api.get('/provinces/closest', {
      params: { latitude, longitude }
    });
    return response.data.data;
  }
};

export default provinceService;

File: backend/src/main/java/com/mapic/backend/services/ProvinceService.java
package com.mapic.backend.services;

import com.mapic.backend.entities.Province;
import com.mapic.backend.repositories.ProvinceRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
@Slf4j
public class ProvinceService {
    
    private final ProvinceRepository provinceRepository;
    
    public List<Province> getAllProvinces() {
        return provinceRepository.findAll();
    }
    
    public Optional<Province> getProvinceByCode(String code) {
        return provinceRepository.findByCode(code);
    }
    
    public Optional<Province> getProvinceByName(String name) {
        return provinceRepository.findByName(name);
    }
    
    public List<Province> getProvincesByRegion(String region) {
        return provinceRepository.findByRegion(region);
    }
    
    public List<Province> searchProvinces(String keyword) {
        return provinceRepository.searchByKeyword(keyword);
    }
    
    /**
     * Find the closest province based on coordinates
     * Uses Haversine formula to calculate distance
     */
    public Optional<Province> findClosestProvince(Double latitude, Double longitude) {
        if (latitude == null || longitude == null) {
            return Optional.empty();
        }
        return provinceRepository.findClosestProvince(latitude, longitude);
    }
    
    /**
     * Detect province from address name
     * Tries to match province name in the address string
     */
    public Optional<Province> detectProvinceFromAddress(String addressName) {
        if (addressName == null || addressName.trim().isEmpty()) {
            return Optional.empty();
        }
        
        // Try to find province by searching in address
        List<Province> allProvinces = provinceRepository.findAll();
        
        for (Province province : allProvinces) {
            if (addressName.toLowerCase().contains(province.getName().toLowerCase()) ||
                addressName.toLowerCase().contains(province.getNameEn().toLowerCase())) {
                return Optional.of(province);
            }
        }
        
        return Optional.empty();
    }
}
